sw/snRuntime/include/dma.h

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+#include <stddef.h>
+#include <stdint.h>
+
+/// A DMA transfer indentifier.
+typedef uint32_t dma_txid_t;
+
+/// Initiate an asynchronous 1D DMA transfer with wide 64-bit pointers.
+static inline dma_txid_t dma_start_1d_wideptr(uint64_t dst, uint64_t src,
+                                              size_t size) {
+    register uint32_t reg_dst_low asm("a0") = dst >> 0;    // 10
+    register uint32_t reg_dst_high asm("a1") = dst >> 32;  // 11
+    register uint32_t reg_src_low asm("a2") = src >> 0;    // 12
+    register uint32_t reg_src_high asm("a3") = src >> 32;  // 13
+    register uint32_t reg_size asm("a4") = size;           // 14
+
+    // dmsrc a0, a1
+    asm volatile(
+        ".word (0b0000000 << 25) | \
+               (     (13) << 20) | \
+               (     (12) << 15) | \
+               (    0b000 << 12) | \
+               (0b0101011 <<  0)   \n" ::"r"(reg_src_high),
+        "r"(reg_src_low));
+
+    // dmdst a0, a1
+    asm volatile(
+        ".word (0b0000001 << 25) | \
+               (     (11) << 20) | \
+               (     (10) << 15) | \
+               (    0b000 << 12) | \
+               (0b0101011 <<  0)   \n" ::"r"(reg_dst_high),
+        "r"(reg_dst_low));
+
+    // dmcpyi a0, a4, 0b00
+    register uint32_t reg_txid asm("a0");  // 10
+    asm volatile(
+        ".word (0b0000010 << 25) | \
+               (  0b00000 << 20) | \
+               (     (14) << 15) | \
+               (    0b000 << 12) | \
+               (     (10) <<  7) | \
+               (0b0101011 <<  0)   \n"
+        : "=r"(reg_txid)
+        : "r"(reg_size));
+
+    return reg_txid;
+}
```

Comment:
My suggestion would be to move all of these functions into a `src/dma.c` file and rely on link-time optimization to inline them at the very end. Otherwise it becomes increasingly difficult to swap in different implementations of the runtime depending on the chosen system (Until now the header files are system-invariant, which would be a nice thing to keep.)

Also, how about we prefix all these functions with `snrt_`, e.g., `snrt_dma_start_1d`, and move all the function prototypes to the single `snrt.h` header? This will keep the include space clean.

Just some nitpicking. What do you think @zarubaf ?

Reply:
No nitpicking at all - this is useful advice and I agree. I've changed the implementation so that the prototypes are now in `snrt.h` and the definitions are in a separate compilation unit (`dma.c`). Agree that this looks much cleaner. I've also seen that LTO is pretty aggressive so I am very positive about this change!